Post Likes: Create a data-layer for post likes

blowery opened the pull request:
Move the existing redux action work into the data layer and add new data layer bits to like and unlike a post.

samouri commented on client/state/data-layer/wpcom/sites/posts/likes/new/test/index.js:
```diff
@@ -0,0 +1,15 @@
+
+describe( 'fromApi', () => {
+	test.skip( 'checks things', () => {
+		//expect( 'ben' ).toEqual( 'b3n' );
```
🚢

alisterscott commented:
I'm just re-adding needs-review label for canary tests

samouri commented on client/state/data-layer/wpcom/sites/posts/likes/index.js:
```diff
@@ -0,0 +1,43 @@
+		apiVersion: '1.1',
+	} );
+
+export const fromApi = data => ( {
```
is there a failure scenario this should throw for?

samouri replied:
....?

samouri commented on client/state/data-layer/wpcom/sites/posts/likes/mine/delete/index.js:
```diff
@@ -0,0 +1,48 @@
+
+export function fromApi( response ) {
+	return {
+		success: !! response.success,
```
instead of passing through a success, i think this should throw in the case of failure.

```js
export function fromApi( response ) {
  if ( ! response || ! response.success ) {
    throw new Error('unsuccesful post unlike');
  }
}
```

blowery replied:
done. rejiggered it quite a bit too.

blowery commented:
in D10183-code we're going to get the liker back from the like and unlike APIs, so added some bits in to expose that. Having the liker will allow the reducers to update the likers array appropriately when the API call returns.

samouri commented on client/state/data-layer/wpcom/sites/posts/likes/mine/delete/index.js:
```diff
@@ -0,0 +1,51 @@
+	);
+
+export const onSuccess = ( { siteId, postId }, { likeCount, liker } ) =>
+	removeLiker( siteId, postId, likeCount, liker );
```
does `liker` always equal the currently active user ?

blowery replied:
right now, yes, but the the shape may be different from the current user in redux.

samouri commented:
Code is looking really clean!
I think this PR is missing the step where you remove the redux-thunk style fetching from `client/state/posts/likes/actions.js`.

blowery commented:
> As-is will this be duping all like requests in both systems?

ah, good call.

blowery commented:
Removed the old fetcher, updated the query component, and a few other places that were using `isRequesting` where they could just be checking for the unloaded state. 

Additional test burden: check the post likes in stats to make sure they still work and appear as you would expect.

samouri commented on client/components/data/query-post-likes/index.jsx:
```diff
@@ -32,25 +30,13 @@ class QueryPostLikes extends Component {
-
-		props.requestPostLikes( siteId, postId );
+	request( { requestPostLikes: rpl, siteId, postId } ) {
+		rpl( siteId, postId );
```
❓ `rpl` --> `requestPostLikes` ❓

blowery replied:
yeah eslint was complaining about shadowing the imported requestPostLikes

samouri commented on client/state/posts/likes/actions.js:
```diff
@@ -22,44 +18,13 @@ import {
-					error,
-				} );
-			} );
+	return {
```
🎉

samouri commented:
testing this out now

dmsnell commented on client/state/data-layer/wpcom/sites/posts/likes/mine/delete/index.js:
```diff
@@ -0,0 +1,51 @@
+export const onError = ( { siteId, postId } ) => bypassDataLayer( like( siteId, postId ) );
+
+export default {
+	[ POST_UNLIKE ]: [
```
I like to encourage keeping with the difference between requests and fact even with `POST`s

We may want to like a post but it may not be possible. We also end up with weird things like `POST_LIKES_REMOVE_LIKER` when `POST_UNLIKE` is probably what we'd rather dispatch in order to have the reducers remove the like.

```js
[ POST_UNLIKE_REQUEST ]: [ dispatchRequestEx( {
  fetch,
  onSuccess: ( { siteId, postId }, { likeCount, liker } ) => [
    { type: POST_UNLIKE, siteId, postId, liker },
    { type: POST_LIKES_SET, siteId, postId, likeCount }
  ]
} ]
```

something like that

dmsnell commented on client/state/data-layer/wpcom/sites/posts/likes/mine/delete/test/index.js:
```diff
@@ -0,0 +1,65 @@
+		const action = fetch( { siteId: 1, postId: 1 } );
+		expect( action ).toHaveProperty( 'method', 'POST' );
+		expect( action ).toHaveProperty( 'path', '/sites/1/posts/1/likes/mine/delete' );
+		expect( action ).toHaveProperty( 'query.apiVersion', '1.1' );
```
alternatively we can also match for conformity

```js
expect( fetch( { siteId: 1, postId: 1 } ) ).toEqual( expect.objectContaining( {
	method: 'POST',
	path: '/sites/1/posts/1/likes/mine/delete',
	query: {
		apiVersion: '1.1'
	}
} ) );
```

dmsnell commented on client/state/data-layer/wpcom/sites/posts/likes/new/index.js:
```diff
@@ -0,0 +1,51 @@
+export const onSuccess = ( { siteId, postId }, { likeCount, liker } ) =>
+	addLiker( siteId, postId, likeCount, liker );
+
+export const onError = ( { siteId, postId } ) => bypassDataLayer( unlike( siteId, postId ) );
```
with `POST_LIKE_REQEUEST` and `POST_UNLIKE_REQUEST` we wouldn't have this condition where we have to bypass the data layer because the data layer is only listening for things that need to sync with the remote

dmsnell commented on client/state/data-layer/wpcom/sites/posts/likes/new/index.js:
```diff
@@ -0,0 +1,51 @@
+		likeCount: +response.like_count,
+		liker: response.liker,
+	};
+}
```
nice!
